storage/namespace_readers.go

```diff
@@ -0,0 +1,313 @@
+	filesetExistsAt(
+		shard uint32,
+		blockStart time.Time,
+	) bool
+
+	get(
+		shard uint32,
+		blockStart time.Time,
+		position readerPosition,
+	) (fs.FileSetReader, error)
+
+	put(reader fs.FileSetReader)
+
+	tick()
+
+	close()
+}
+
+type namespaceReaderManager struct {
+	sync.Mutex
+
+	filesetExistsAtFn fsFilesetExistsAtFn
+	newReaderFn       fsNewReaderFn
+
+	namespace namespace.Metadata
+	fsOpts    fs.Options
+	bytesPool pool.CheckedBytesPool
+
+	logger xlog.Logger
+
+	closedReaders []cachedReader
+	openReaders   map[cachedOpenReaderKey]cachedReader
+
+	metrics namespaceReaderManagerMetrics
+}
+
+type cachedOpenReaderKey struct {
+	shard      uint32
+	blockStart xtime.UnixNano
+	position   readerPosition
+}
+
+type readerPosition struct {
+	dataIdx     int
+	metadataIdx int
+}
+
+type cachedReader struct {
+	reader         fs.FileSetReader
+	ticksSinceUsed int
+}
+
+type namespaceReaderManagerMetrics struct {
+	cacheHit              tally.Counter
+	cacheMissAllocReader  tally.Counter
+	cacheMissReusedReader tally.Counter
+}
+
+func newNamespaceReaderManagerMetrics(
+	scope tally.Scope,
+) namespaceReaderManagerMetrics {
+	subScope := scope.SubScope("reader-cache")
+	return namespaceReaderManagerMetrics{
+		cacheHit: subScope.Counter("hit"),
+		cacheMissAllocReader: subScope.Tagged(map[string]string{
+			"miss_type": "alloc_reader",
+		}).Counter("miss"),
+		cacheMissReusedReader: subScope.Tagged(map[string]string{
+			"miss_type": "reuse_reader",
+		}).Counter("miss"),
+	}
+}
+
+func newNamespaceReaderManager(
+	namespace namespace.Metadata,
+	namespaceScope tally.Scope,
+	opts Options,
+) databaseNamespaceReaderManager {
+	return &namespaceReaderManager{
+		filesetExistsAtFn: fs.FilesetExistsAt,
+		newReaderFn:       fs.NewReader,
+		namespace:         namespace,
+		fsOpts:            opts.CommitLogOptions().FilesystemOptions(),
+		bytesPool:         opts.BytesPool(),
+		logger:            opts.InstrumentOptions().Logger(),
+		openReaders:       make(map[cachedOpenReaderKey]cachedReader),
+		metrics:           newNamespaceReaderManagerMetrics(namespaceScope),
+	}
+}
+
+func (m *namespaceReaderManager) filesetExistsAt(
+	shard uint32,
+	blockStart time.Time,
+) bool {
+	return m.filesetExistsAtFn(m.fsOpts.FilePathPrefix(),
+		m.namespace.ID(), shard, blockStart)
+}
+
+func (m *namespaceReaderManager) get(
+	shard uint32,
+	blockStart time.Time,
+	position readerPosition,
+) (fs.FileSetReader, error) {
+	m.Lock()
+
+	key := cachedOpenReaderKey{
+		shard:      shard,
+		blockStart: xtime.ToUnixNano(blockStart),
+		position:   position,
+	}
+
+	openReader, ok := m.openReaders[key]
+	if ok {
+		// Cache hit, take this open reader
+		delete(m.openReaders, key)
+		m.Unlock()
+
+		m.metrics.cacheHit.Inc(1)
+		return openReader.reader, nil
+	}
+
+	// Cache miss, need to return a reused reader or open a new reader
+	var (
+		reader fs.FileSetReader
+		err    error
+	)
+	if len(m.closedReaders) > 0 {
+		idx := len(m.closedReaders) - 1
```

Comment:
super nit: might be nice to wrap these lines into a pop() method and the append(m.closedReader, ...) in put into a push() method. That way its much easier to realize that you're treating m.closedReaders as a stack

Reply:
Done, ta.